chall09/gbudau.c

```diff
@@ -0,0 +1,164 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+#include <ctype.h>
+#include <errno.h>
+#define MAX_SECONDS 10000000000000000000UL
+#define YEAR 31536000UL
+#define DAY 86400UL
+#define HOUR 3600UL
+#define MINUTE 60UL
+#define ARRAY_SIZE 5
+
+static const char		*unit_time[ARRAY_SIZE] = {"year", "day", "hour", "minute", "second"};
+
+static char				*walk_space(char *seconds)
+{
+	while (*seconds && *seconds == ' ')
+		seconds++;
+	return (seconds);
+}
+
+static unsigned long	set_error(unsigned long ret, int *err)
+{
+	*err = 1;
+	return (ret);
+}
+
+static unsigned long	get_seconds(char *seconds, int *err)
+{
+	char			*end;
+	unsigned long	ul_seconds;
+
+	*err = 0;
+	seconds = walk_space(seconds);
+	if (*seconds == '\0' || isdigit(*seconds) == 0)
+		return (set_error(0, err));
+	errno = 0;
+	ul_seconds = strtoul(seconds, &end, 10);
+	if (errno == ERANGE)
+		return (set_error(0, err));
+	if (*end != '\0')
+	{
+		end = walk_space(end);
+		if (*end != '\0')
+			return (set_error(0, err));
+	}
+	return (ul_seconds);
+}
+
+static unsigned			unit_length(unsigned long unit)
+{
+	unsigned	length = 0;
+
+	while (unit > 0)
+	{
+		length++;
+		unit /= 10;
+	}
+	return (length);
+}
+
+static unsigned			total_length(unsigned long time[], unsigned *positive)
+{
+	unsigned				total_len = 0;
+
+	*positive = 0;
+	for (int i = 0; i < ARRAY_SIZE; i++)
+	{
+		if (time[i] > 0)
+		{
+			(*positive)++;
+			total_len += unit_length(time[i]) + 1 + strlen(unit_time[i]) + (time[i] > 1);
+		}
+	}
+	return (total_len);
+}
+
+static unsigned			separator_length(unsigned positive)
+{
+	if (positive == 1)
+		return (0);
+	return (strlen(", ") * (positive - 2) + strlen(" and "));
+}
+
+static char				*free_fmt(char *format)
+{
+	if (format != NULL)
+		free(format);
+	return (format = NULL);
```

Comment:
There's no use in doing `format = NULL` here. Format won't ever be used again as you exit de function.